pkg/util/awsutil/s3_client.go

```diff
@@ -0,0 +1,117 @@
+// Copyright 2017 The etcd-operator Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package awsutil
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+
+	api "github.com/coreos/etcd-operator/pkg/apis/etcd/v1beta2"
+
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	tmpdir = "/tmp"
+)
+
+// S3ClientFactory can be used to generate S3 clients from e.g. given aws secret.
+// It helps handle details around k8s logic and temp dir.
+type S3ClientFactory struct {
+	kubecli   kubernetes.Interface
+	namespace string
+	awsSecret string
+}
+
+// S3Client is a wrapper for S3 client that provides cleanup functionality.
+type S3Client struct {
+	S3        *s3.S3
+	configDir string
+}
+
+func NewS3ClientFactory(kubecli kubernetes.Interface, namespace, awsSecret string) *S3ClientFactory {
+	return &S3ClientFactory{
+		kubecli:   kubecli,
+		namespace: namespace,
+		awsSecret: awsSecret,
+	}
+}
+
+// NewClient returns a S3 client.
+func (f *S3ClientFactory) NewClient() (w *S3Client, err error) {
```

Comment:
Why can't we just have `NewS3Client` as the factory method to create the `S3Client`?  e.g:
```go
func NewS3Client(kubecli kubernetes.Interface, namespace, awsSecret string)  (w *S3Client, err error) {
...
}
```